GoogleForCreators/web-stories-wp
# Work around KSES for users without unfiltered_html capability

## Summary

This PR aims to solve the critical KSES bug for users without the `unfiltered_html` capability (authors on single site / editors and site admins on multisite).

The solution is loosely based on the one we came up with for the legacy stories editor in https://github.com/ampproject/amp-wp/pull/2691.

## Relevant Technical Choices

* Fully disables KSES for `post_content_filtered`, where we story our story data as a JSON blob. This JSON data is only ever used in the editor, and isn't directly displayed on the frontend or anything. Since we're in full control of it, it should be safe to remove.  
  Related core ticket: https://core.trac.wordpress.org/ticket/38715
* Copies over `safecss_filter_attr` from core with two big modifications:  
    1. Extends the list of allowed elements and attributes for KSES, and the list of allowed CSS properties so that our stories remain valid and unchanged.  
    **Note:** The extension is done via the `safe_style_css` filter so it's easier to maintain.
    1. Adds support for `transform` CSS property. Otherwise this would prevent inline styles like `transform: rotate(90deg);` because the function strips any parentheses.  
    **Note:** This could potentially be contributed upstream.
    1. Adds support for `rgb(a)` usage for colors. Otherwise this would prevent things like adding opacity to colors.  
    **Note:** This could potentially be contributed upstream (https://core.trac.wordpress.org/ticket/24157).
* Only applies these changes for the REST API context, to minimize risk affecting other parts of the site.
* Adds new `Stories_Autosaves_Controller` to make sure KSES changes also apply when creating autosaves (i.e. when previewing a published story)

## To-do

* [ ] Field Testing
* [ ] Investigate failing E2E tests on Firefox Nightly

## User-facing changes



## Testing Instructions

1. Log in as user with role Author
1. Create a new story
1. Add a new text element
1. Make the text bold
1. Rotate element
1. Add background color with opacity
1. Preview story
1. Verify story is displayed as expected

---



Fixes #1121
